Allow e.com to link straight to e.cash and share sessions

tgolen opened the pull request:
### Fixed Issues
Done along with: https://github.com/Expensify/Web-Expensify/pull/31302
Fixes https://github.com/Expensify/Expensify/issues/162554

### Tests / QA
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITHOUT two-factor authentication
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
4. Verify that you are automatically logged into e.cash and are seeing the page to create a new workspace

Now, test it out with 2FA

1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITH two-factor authentication enabled
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
4. Verify that you are redirected to e.cash to ever a 2FA code
5. After the code is entered, verify you are automatically logged into e.cash and are seeing the page to create a new workspace

### Tested On
No other platforms need to be tested because this flow only works on web (the inbox task isn't shown on mobile at all)
- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
![image](https://user-images.githubusercontent.com/1228807/124313390-0d1e4c00-db2e-11eb-9c85-abe239b54817.png)

cc @marcaaron @TomatoToaster

tgolen committed:
Add new validation routes

tgolen committed:
Use a constant for the screen names

tgolen committed:
Add the route to public screens as well

tgolen committed:
Add a comment for the route

tgolen committed:
Move variable to a class property

tgolen committed:
Create separate pages for the separate routes

tgolen committed:
WIP on redirecting to new workspace page

tgolen committed:
Get redirection to the new workspace modal working

tgolen committed:
Add a user action for getting an authToken

tgolen committed:
Lint and show proper error message

tgolen committed:
Rename action method

tgolen committed:
Move method to Session action and implement API method

tgolen committed:
Pass the 2FA code to action method

tgolen committed:
Remove unneeded data from request

tgolen committed:
Merge branch 'main' into tgolen-validate-redirect

tgolen committed:
Improve comment, redirect in the action file

tgolen committed:
Merge branch 'main' into tgolen-validate-redirect

tgolen commented:
This is all working, but the redirect to the new-workspace page is causing several JS console warnings:

![image](https://user-images.githubusercontent.com/1228807/124313602-4fe02400-db2e-11eb-9059-2f65123ff84f.png)

They don't appear to be affecting anything, but I haven't been able to figure out what the deal is. My guess is that it's related to running these back-to-back:

```
            Navigation.dismissModal();
            Navigation.navigate(ROUTES.WORKSPACE_NEW);
```

and I think there is a race condition because I think part of the code inside `dismissModal()` is async, but it's hard to figure out when exactly `dismissModal()` is finished doing its thing. There is one existing flow that uses this same dismiss->navigate pattern here: https://github.com/Expensify/Expensify.cash/blob/099457ec41f238be8103d953e4c8b50ea2b030f1/src/libs/actions/Policy.js#L201-L202

I'm not sure what to do about this. We need to have this code deployed by Jul 9, so I am thinking that we merge this as-is, and then open up an issue for improving this dismiss->navigate pattern.

marcaaron reviewed (commented):
These changes LGTM! I'm not sure whether there is any problem with the dismiss + navigate pattern (it seems kosher to me). But I agree that the react warning is not a blocker. Could be an issue is `withOnyx()` not having any awareness that its children have unmounted - I've observed similar issues before, but hard to say without debugging.

tgolen commented on src/pages/ValidateLogin2FANewWorkspacePage.js:
```diff
@@ -0,0 +1,152 @@
+        if (this.props.session.authToken) {
+            // In order to navigate to a modal, we first have to dismiss the current modal. But there is no current
+            // modal you say? I know, it confuses me too. Without dismissing the current modal, if they user cancels
+            // out of the new workspace modal, then they will be routed back to
```
Yeah, I'm all in favor of updating this comment to make more sense. If the real problem is that `ValidateLogin2FANewWorkspacePage` is in the history stack, maybe there is a better way for us to remove it from the stack without needing to call `dismissModal()`?

marcaaron replied:
>maybe there is a better way for us to remove it from the stack without needing to call dismissModal()

I'm having trouble working out a better way. It feels like there are contexts where you'd want the default `navigate()` method to preserve the underlying modal in the stack e.g. moving from `/settings` to `/settings/payments` and contexts where you'll want a different behavior. We could maybe add a parameter to `navigate()` that forces all other modals to close?

tgolen replied:
All right, I was able to fix this. Turns out, I was navigating to that modal twice (once in Session, and again from ComponentDidMount (from when the same page was remounted: once in `public screens` and then again in `auth screens`).

Once I removed the navigation in the Session action, it all worked like a charm.

tgolen commented:
I removed the hold because this actually needs to be deployed first, before the web-e change.

alex-mechler commented on src/languages/en.js:
```diff
@@ -267,6 +267,7 @@ export default {
     },
     passwordForm: {
         pleaseFillOutAllFields: 'Please fill out all fields',
+        EnterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```
```suggestion
        enterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```

alex-mechler commented on src/pages/ValidateLogin2FANewWorkspacePage.js:
```diff
@@ -0,0 +1,152 @@
+            return;
+        }
+
+
```
Extra blank line here

tgolen committed:
Remove extra call to continue session

tgolen committed:
Merge branch 'main' into tgolen-validate-redirect

tgolen committed:
Remove extra navigation

tgolen committed:
Fix prop types

tgolen committed:
Lower case the translate key

tgolen committed:
Remove extra space

tgolen committed:
Upper case API method.

tgolen commented:
All right, thanks for the reviews! I updated it and addressed everything as well as fixing the initial warnings that I was struggling with. Should be ready for another review.

marcaaron commented on src/libs/API.js:
```diff
@@ -296,6 +295,33 @@ function reauthenticate(command = '') {
+ * @param {String} parameters.accountID
+ * @param {String} parameters.validateCode
+ * @param {String} [parameters.twoFactorAuthCode]
+ * @returns {Promise<unknown>}
```
NAB / just curious but why `Promise<unknown>`?

tgolen replied:
I think PHPStorm did that automatically, so I'm not sure why! I'll remove it

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -93,6 +95,14 @@ const modalScreenListeners = {
+    },
+    headerShown: false,
+    animationTypeForReplace: 'pop',
+};
```
Seems like `defaultScreenOptions` is [also declared in `PublicScreens.js` with the same values](https://github.com/Expensify/Expensify.cash/blob/36a336400efc060df2ebc21d74ca1ee19ee5afff/src/libs/Navigation/AppNavigator/PublicScreens.js#L10-L16). Maybe they should be moved into a separate config file so we don't have to remember to update things twice in the future when applying a change?

tgolen replied:
Sure, I can do that.

tgolen committed:
Make default screen options into a reusable module

tgolen commented:
Updated to DRY up the default screen options.

alex-mechler commented:
I'm trying to test this out, and Web-E redirects me to the correct link, but then I am redirect to the homepage without any error messages. Is there any additional setup I have to do that is not mentioned in the testing steps?

tgolen commented:
@alex-mechler I don't think there is anything special you need to do, no.

Just so I am clear, the test steps work for except that instead of seeing the new-workspace page, you see the home page?

alex-mechler commented:
Correct. I am not automatically logged into e.cash, and do not see the new workspace page. I checked out the branch from https://github.com/Expensify/Web-Expensify/pull/31302 and this branch. Is there any other branches I need to checkout? Was there any auth changes that I might be missing?

alex-mechler commented on src/languages/en.js:
```diff
@@ -268,6 +268,7 @@ export default {
     },
     passwordForm: {
         pleaseFillOutAllFields: 'Please fill out all fields',
+        enterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```
NAB: I feel like `your` reads better than `a` here

tgolen replied:
OK, I can change this.

alex-mechler commented on src/pages/ValidateLogin2FANewWorkspacePage.js:
```diff
@@ -0,0 +1,146 @@
+                            success
+                            style={[styles.mb2]}
+                            text={this.props.translate('common.continue')}
+                            isLoading={this.state.loading}
```
NAB: We never set this anywhere but the constructor, meaning the user has no feedback as to when we are actually loading

tgolen replied:
I'll see about cleaning this up.

alex-mechler reviewed (dismissed):
Turns out my auth was out of date 😅 

Works well, two NABs, feel free to merge if you don't think they are worth fixing

tgolen committed:
Reword form prompt

tgolen committed:
Set the loading state to true

tgolen commented:
OK, I updated to change the language on the form prompt and to display the loading indicator when the 2FA is being validated.

tgolen commented:
I'm gonna go ahead and merge since the other reviewers approved already and the changes since then were minor.

tgolen commented:
Once this goes to production, we can merge Expensify/Web-Expensify#31302

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

kavimuru commented:
##  User is not automatically logged into e.cash after completing the inbox task
## Expected Result: 
Scenario 1: ​Verify that you are automatically logged into e.cash and are seeing the page to create a new workspace
Scenario 2:  Verify that you are redirected to e.cash to ever a 2FA code
## Actual Result: 
Not automatically logged into e.cash and the page to create a new workspace isn't shown. User is redirected to sign in page for both the scenarios
## Actions Performed:
Scenario1:
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITHOUT two-factor authentication
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card

Scenario2:
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITH two-factor authentication enabled
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
## Platform:
Web ✔️
iOS 
mWeb 
Android 
## Build:
1.0.77-0
## Notes/Images/Video:
![Bug5147298_Screen_Shot_2021-07-10_at_7 30 21_PM](https://user-images.githubusercontent.com/43996225/125181813-8e34ad80-e1d6-11eb-8ca1-f811d7e94375.png)

![Bug5147298_Screen_Shot_2021-07-10_at_7 35 37_PM](https://user-images.githubusercontent.com/43996225/125181815-8f65da80-e1d6-11eb-813e-f9dbb68f8764.png)

tgolen commented:
Ah, I think that's expected for now. Sorry, this issue should be production QA and internal QA (because it requires being in a beta for it to work). It also needs to have https://github.com/Expensify/Web-Expensify/pull/31302 deployed, and it looks like that still hasn't been.

isagoico commented:
Perfect! will check the PR off the list then. Thanks 🙇

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/1031429034) to production         in version: 1.0.77-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

TomatoToaster commented:
ITS WORKING!!
I got redirected from expensify after gettting the new task
![image](https://user-images.githubusercontent.com/19364431/125834154-2c196589-a829-4269-8d6b-f7fcf87e99f6.png)

tgolen commented:
Yeah! Noice

On Thu, Jul 15, 2021 at 12:09 PM Amal Nazeem ***@***.***>
wrote:

> ITS WORKING!!
> I got redirected from expensify after gettting the new task
> [image: image]
> <https://user-images.githubusercontent.com/19364431/125834154-2c196589-a829-4269-8d6b-f7fcf87e99f6.png>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/Expensify.cash/pull/3715#issuecomment-880907808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJMAB2LEGKRHYBFGDHV7O3TX4P4NANCNFSM47CSENHA>
> .
>
